### PR TITLE
Adds capability to sample variables at the location of a ray

### LINKED ISSUE #12: MOOSE Variable Sampling at a Ray Location
## Reason
Charged particles are subject to the Lorentz force. Since the electric and magnetic fields will be MOOSE Variables we need a way to sample the value of a field at the point a ray currently exists. 

## Design
A utility will be used to give the user a simple interface to sample variables at the location of a ray as the ray moves through space. 

## Impact
This will allow the user to calculate the Lorentz force on a charged particle as it moves through space.

### PR DESCRIPTION
**Pull request by gsgall**
Closes #12

### CODE REVIEW

File: doc/content/source/utils/VariableSampler.md [hunk 1]
```diff
@@ -0,0 +1,5 @@
+# VariableSampler
+
+The `VariableSampler` is a utility that enables users to sample the value of a MooseVariable at an abitrary point in space
+
+The sampler is designed to be used with rays. The point parameter exist within the provided element parameter. After instantiation, use `VariableSampler::sampleVariable()` to sample the value of a variable at a point in space.
```
Comment (cticenhour) [inline doc/content/source/utils/VariableSampler.md]:
```suggestion
The sampler is designed to be used with rays. The point parameter exists within the provided element parameter. After instantiation, use `VariableSampler::sampleVariable()` to sample the value of a variable at a point in space.
```

File: include/utils/VariableSampler.h [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#pragma once
+
+#include "FEProblemBase.h"
+#include "libmesh/system.h"
+
+namespace FENIX
+{
+/**
+ * Utility that samples the value of a vector variable at a given point
```
Comment (cticenhour) [inline include/utils/VariableSampler.h]: You mention vector variables explicitly here, but not in the documentation markdown. Is this for all variables or only vectors? Depending on which it is, the markdown or this line should be updated to reflect that.

File: include/utils/VariableSampler.h [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+ */
+class VariableSampler
+{
+public:
+  VariableSampler(FEProblemBase & problem, const VariableName & name, const THREAD_ID tid);
+
+  /**
+   *  Method for getting the value of a variable at a given point
+   * @param point the point where you want to know the value of a variable
```
Comment (cticenhour) [inline include/utils/VariableSampler.h]:
```suggestion
   * @param point the point at which you want to know the value of a variable
```

File: include/utils/VariableSampler.h [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+   * @param element the element in which the point you want to sample resides
+   * @returns the value of the variable of interest at the point of interest
+   */
+  Real sampleVariable(const Point & point, const Elem * element) const;
+
+private:
+  /// A reference to the system containing the variable
+  const System & _system;
+  /// the variable number for the vector field in the system
```
Comment (cticenhour) [inline include/utils/VariableSampler.h]:
```suggestion
  /// The variable number for the vector field in the system
```

File: test/include/userobjects/TestVariableSampler.h [hunk 5]
```diff
@@ -0,0 +1,20 @@
+#pragma once
+
+#include "VariableSampler.h"
+#include "TestPICStudyUserDefinedRays.h"
+class TestVariableSampler : public TestPICStudyUserDefinedRays
```
Comment (cticenhour) [inline test/include/userobjects/TestVariableSampler.h]:
```suggestion
#include "TestPICStudyUserDefinedRays.h"

class TestVariableSampler : public TestPICStudyUserDefinedRays
```

File: test/include/userobjects/TestVariableSampler.h [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+
+  static InputParameters validParams();
+
+protected:
+  virtual void postExecuteStudy() override;
+
+private:
+  /// the utility used to sample a moose variable as a ray is traveling
+  const FENIX::VariableSampler _sampler;
+  /// the ray data index where the value of sampled field will be stored
```
Comment (cticenhour) [inline test/include/userobjects/TestVariableSampler.h]:
```suggestion
  /// The utility used to sample a moose variable as a ray is traveling
  const FENIX::VariableSampler _sampler;
  /// The ray data index where the value of sampled field will be stored
```

File: test/tests/utils/variable_sampler/tests [hunk 7]
```diff
@@ -0,0 +1,21 @@
...
+               variable_sampler_rays.e-s0002
+               variable_sampler_rays.e-s0003
+               variable_sampler_rays.e-s0004
+               variable_sampler_rays.e-s0005
+               variable_sampler_rays.e-s0006
+               variable_sampler_rays.e-s0007
+               variable_sampler_rays.e-s0008
+               variable_sampler_rays.e-s0009'
+    allow_test_objects = true
+    requirement = 'A ray should be able to sample the value of field variable at any point in space as it moves through space'
```
Comment (cticenhour) [inline test/tests/utils/variable_sampler/tests]:
```suggestion
    requirement = 'The system shall allow a ray to sample the value of field variable at any point in space as it moves through space'
```

**Review by cticenhour — CHANGES_REQUESTED** [review]
Beyond these noted comments, please also change the title of the PR to be more reflective of the PR content and change the commit message to reflect the content (having `closes #12` at the end of the commit is still OK - we just need a descriptive commit message).

File: include/utils/VariableSampler.h (hunk 2 above)
Comment (gsgall) [reply include/utils/VariableSampler.h]: That is left over from the original one where I was trying to sample vectors but from our discussions in slack this does not work for vector variables. I have updated this comment to reflect the real ability of this utility.

**Review by cticenhour — APPROVED** [review]

**Pull request merged**